File: frontend/src/game/upgrades/stat_modifiers/damage_3.ts
import { Upgrade, type UpgradeDef } from '../Upgrade'
import { UpgradeTargetID, UpgradeStatID, RarityID, UpgradeTypeID } from '../../data/ID'

export const Damage3Def: UpgradeDef = {
  id: "damage_3",
  name: "Devastation",
  description: "+1.6% damage.",
  rarity: RarityID.Rare,
  upgradeType: UpgradeTypeID.StatModifier,
  cost: 10,
  targetClass: UpgradeTargetID.Attack,
  fieldInTargetClass: UpgradeStatID.Damage,
  value: 0.016,
  isMultiplier: true,
  stackable: true,
  maxStacks: 99999,
}

export class Damage3 extends Upgrade {}

File: backend/app/core/upgrades/stat_modifiers/damage_3.py
from ..upgrade_implementation import UpgradeImplementation, UpgradeAppliers


class Damage3(UpgradeImplementation):
    id = 'damage_3'
    name = 'Devastation'
    description = '+1.6% damage.'
    rarity = 'rare'
    type = 'stat_modifier'
    cost = 10

    def can_apply(self, state: dict) -> bool:
        return True

    def apply(self, appliers: UpgradeAppliers) -> None:
        value = 0.016
        appliers.modifiers.add_multiplier('attack', 'damage', value)

    def validate(self, recorded_stats: dict) -> bool:
        return True
